### PR TITLE
Integrate project reactor

### PR DESCRIPTION
**Pull request by ajax-yurkovskyi-t**

### CODE REVIEW

File: src/main/kotlin/com/example/iotmanagementdevice/security/AuthenticationServiceImpl.kt [hunk 1]
```diff
@@ -17,20 +17,20 @@ class AuthenticationServiceImpl(
...
-
+        }.map { authentication ->
             val securityUser = authentication.principal as SecurityUser
             val token = jwtUtil.generateToken(securityUser.username)
             UserLoginResponseDto(token)
-        } catch (ex: BadCredentialsException) {
-            throw AuthenticationException(LOGIN_ERROR_MESSAGE, ex)
+        }.onErrorMap(BadCredentialsException::class.java) { ex ->
+            AuthenticationException(LOGIN_ERROR_MESSAGE, ex)
         }
```
Comment (ajax-torbin-v) [inline src/main/kotlin/com/example/iotmanagementdevice/security/AuthenticationServiceImpl.kt:34]: Why do you don't create your error lazily?

File: src/main/kotlin/com/example/iotmanagementdevice/security/CustomUserDetailsService.kt [hunk 2]
```diff
@@ -13,7 +13,7 @@ class CustomUserDetailsService(
 ) : UserDetailsService {
 
     override fun loadUserByUsername(email: String): UserDetails? {
-        val mongoUser = userRepository.findByUserEmail(email)
+        val mongoUser = userRepository.findByUserEmail(email).block()
```
Comment (ajax-torbin-v) [inline src/main/kotlin/com/example/iotmanagementdevice/security/CustomUserDetailsService.kt:16]: ?????????
🤯

File: src/main/kotlin/com/example/iotmanagementdevice/repository/UserRepositoryImpl.kt [hunk 3]
```diff
@@ -41,33 +43,36 @@ class UserRepositoryImpl(private val mongoTemplate: MongoTemplate) :
             Update().set(MongoDevice::userId.name, ObjectId(userId)),
             MongoDevice::class.java
         )
-        return userUpdateResult.modifiedCount > 0 && deviceUpdateResult.modifiedCount > 0
+        return Mono.zip(userUpdateResult, deviceUpdateResult)
+            .map { tuple ->
+                tuple.t1.modifiedCount > 0 && tuple.t2.modifiedCount > 0
+            }
```
Comment (ajax-butriy-v) [inline src/main/kotlin/com/example/iotmanagementdevice/repository/UserRepositoryImpl.kt:49]: use Kotlin extensions

File: src/main/kotlin/com/example/iotmanagementdevice/service/user/UserService.kt [hunk 4]
```diff
@@ -4,19 +4,21 @@ import com.example.iotmanagementdevice.dto.device.response.DeviceResponseDto
 import com.example.iotmanagementdevice.dto.user.request.UserRegistrationRequestDto
 import com.example.iotmanagementdevice.dto.user.request.UserUpdateRequestDto
 import com.example.iotmanagementdevice.dto.user.response.UserResponseDto
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
 
 interface UserService {
-    fun register(requestDto: UserRegistrationRequestDto): UserResponseDto
+    fun register(requestDto: UserRegistrationRequestDto): Mono<UserResponseDto> // Changed to Mono
```
Comment (ajax-torbin-v) [inline src/main/kotlin/com/example/iotmanagementdevice/service/user/UserService.kt:11]:
```suggestion
    fun register(requestDto: UserRegistrationRequestDto): Mono<UserResponseDto>
```

File: src/main/kotlin/com/example/iotmanagementdevice/security/CustomUserDetailsService.kt (hunk 2 above)
Comment (ajax-butriy-v) [inline src/main/kotlin/com/example/iotmanagementdevice/security/CustomUserDetailsService.kt:16]: https://docs.spring.io/spring-security/reference/reactive/index.html

File: src/test/kotlin/com/example/iotmanagementdevice/repository/UserRepositoryImplTest.kt [hunk 5]
```diff
@@ -81,23 +82,28 @@ class UserRepositoryImplTest : AbstractMongoTest {
...
-        val users = userRepositoryImpl.findAll()
+        val users = userRepositoryImpl.findAll().collectList()
 
         // Then
-        val expectedUsers = listOf(user1, user2)
-        assertTrue(users.containsAll(expectedUsers), "Expected users not found in the repository")
+        users.test()
+            .expectNextMatches {
+                it.contains(user1) && it.contains(user2)
+            }
```
Comment (ajax-torbin-v) [inline src/test/kotlin/com/example/iotmanagementdevice/repository/UserRepositoryImplTest.kt:105]: .containsAll

File: src/test/kotlin/com/example/iotmanagementdevice/repository/UserRepositoryImplTest.kt [hunk 6]
```diff
@@ -138,69 +145,79 @@ class UserRepositoryImplTest : AbstractMongoTest {
...
-        assertEquals(updatedUserDto.email, foundUser?.email)
-        assertEquals(updatedUserDto.phoneNumber, foundUser?.phoneNumber)
-        assertEquals(updatedUserDto.userPassword, foundUser?.userPassword)
+        userRepositoryImpl.findById(user.id!!.toString()).test()
+            .expectNextMatches {
+                it.name == updatedUserDto.name &&
+                    it.email == updatedUserDto.email &&
+                    it.phoneNumber == updatedUserDto.phoneNumber &&
+                    it.userPassword == updatedUserDto.userPassword
+            }
```
Comment (ajax-torbin-v) [inline src/test/kotlin/com/example/iotmanagementdevice/repository/UserRepositoryImplTest.kt:157]: ?????
Map user to DTO and compare then

File: src/test/kotlin/com/example/iotmanagementdevice/service/user/UserServiceImplTest.kt [hunk 7]
```diff
@@ -336,13 +330,20 @@ class UserServiceImplTest {
...
+            )
+        } returns true.toMono()
 
         // When
         val result = userService.assignDeviceToUser(userObjectId.toString(), deviceObjectId.toString())
 
         // Then
-        assertTrue(result, "Expected the device to be assigned to the user successfully")
+        result.test()
+            .expectNext(true)
```
Comment (ajax-torbin-v) [inline src/test/kotlin/com/example/iotmanagementdevice/service/user/UserServiceImplTest.kt:344]: Add message for true/false assertions pls

File: src/test/resources/application-test.yaml [hunk 8]
```diff
@@ -1,3 +1,6 @@
+spring:
+  main:
+    web-application-type: reactive
```
Comment (ajax-samondros-b) [inline src/test/resources/application-test.yaml:3]: should not it be in application.yaml?

**Review by ajax-samondros-b — APPROVED** [review]

**Pull request merged**